internal/auth/authn/aapgateway.go

```diff
@@ -0,0 +1,104 @@
+package authn
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/flightctl/flightctl/internal/auth/common"
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jwt"
+)
+
+const (
+	AAP_GATEWAY_AUTH_HEADER = "X-Dab-Jw-Token"
+)
+
+type AapGatewayAuth struct {
+	GatewayUrl      string
+	PublicKey       any
+	ClientTlsConfig *tls.Config
+}
+
+func NewAapGatewayAuth(gatewayUrl string, clientTlsConfig *tls.Config) (AapGatewayAuth, error) {
+	aapGatewayAuth := AapGatewayAuth{
+		GatewayUrl:      gatewayUrl,
+		ClientTlsConfig: clientTlsConfig,
+	}
+	client := &http.Client{Transport: &http.Transport{
+		TLSClientConfig: clientTlsConfig,
+	}}
+	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("%s/v1/jwt_key/", gatewayUrl), nil)
+	if err != nil {
+		return aapGatewayAuth, err
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		return aapGatewayAuth, err
+	}
+
+	defer resp.Body.Close()
+	publicKeyPem, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return aapGatewayAuth, err
+	}
+
+	block, _ := pem.Decode(publicKeyPem)
+	if block == nil || block.Type != "PUBLIC KEY" {
+		return aapGatewayAuth, fmt.Errorf("error decoding public key")
+	}
+
+	publicKey, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return aapGatewayAuth, fmt.Errorf("error parsing public key: %w", err)
+	}
+
+	aapGatewayAuth.PublicKey = publicKey
+	return aapGatewayAuth, nil
+}
+
+func (a AapGatewayAuth) ValidateToken(ctx context.Context, token string) (bool, error) {
+	if strings.HasPrefix(token, "Bearer ") {
+		// oauth2 token
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf("%s/v1/me/", a.GatewayUrl), nil)
+		if err != nil {
+			return false, err
+		}
+		req.Header = map[string][]string{
+			"Authorization": {token},
+			"Content-Type":  {"application/json"},
+		}
+
+		client := &http.Client{Transport: &http.Transport{
+			TLSClientConfig: a.ClientTlsConfig,
+		}}
+		res, err := client.Do(req)
+		if err != nil {
+			return false, err
+		}
+		return res.StatusCode == http.StatusOK, nil
+	} else {
+		_, err := jwt.Parse([]byte(token), jwt.WithKey(jwa.RS256, a.PublicKey))
+		return err == nil, err
```

Comment:
We probably want the middleware to store the scopes/claims in the request's ctx, so we can later do the authorization off of these.

Reply:
service layer would just call authz.CheckPermission() and the proper AuthZ implementation will do the check. How the checking is done is abstracted from service layer.

Here is an example usage https://github.com/flightctl/flightctl/blob/main/internal/service/device.go#L50

so if there is 
- k8s AuthZ, it would check agaist the k8s API
- AAP AuthZ would match scopes/claims
- c.rh.c AuthZ would again match some scopes/claims